marlowe/src/Language/Marlowe/Semantics.hs

```diff
@@ -107,108 +156,268 @@ data Observation = AndObs Observation Observation
+                      | TEApplyNoMatchError
+                      | TEIntervalError IntervalError
+                      | TEUselessTransaction
+  deriving (Show)
+
+
+{-| Marlowe transaction input.
+-}
+data TransactionInput = TransactionInput
+    { txInterval :: SlotInterval
+    , txInputs   :: [Input] }
+  deriving (Show)
+
+
+{-| Marlowe transaction output.
+-}
+data TransactionOutput =
+    TransactionOutput
+        { txOutWarnings :: [TransactionWarning]
+        , txOutPayments :: [Payment]
+        , txOutState    :: State
+        , txOutContract :: Contract }
+    | Error TransactionError
+  deriving (Show)
 
 
+{-|
+    This data type is a content of a contract's /Data Script/
+-}
+data MarloweData = MarloweData {
+        marloweCreator  :: Party,
+        marloweState    :: State,
+        marloweContract :: Contract
+    } deriving (Show)
+
+
+-- | Empty State for a given minimal 'Slot'
+emptyState :: Slot -> State
+emptyState sn = State
+    { accounts = Map.empty ()
+    , choices  = Map.empty ()
+    , boundValues = Map.empty ()
+    , minSlot = sn }
+
+-- | Returns an owner of an account
+accountOwner :: AccountId -> Party
+accountOwner (AccountId _ party) = party
+
+
+-- | Check if a 'num' is withint a list of inclusive bounds.
+inBounds :: ChosenNum -> [Bound] -> Bool
+inBounds num = any (\(Bound l u) -> num >= l && num <= u)
+
+
+{- Checks 'interval' and trim it if necessary. -}
 fixInterval :: SlotInterval -> State -> IntervalResult
-fixInterval interval state = let
-    SlotInterval low high  = interval
-    curMinSlot = minSlot state
-    -- newLow is both new "low" and new "minSlot" (the lower bound for slotNum)
-    newLow = max low curMinSlot
-    -- We know high is greater or equal than newLow (prove)
-    curInterval = SlotInterval newLow high
-    env = Environment { slotInterval = curInterval }
-    newState = state { minSlot = newLow }
-    in if high < low then IntervalError (InvalidInterval interval)
-    else if high < curMinSlot then IntervalError (IntervalInPastError curMinSlot interval)
-    else IntervalTrimmed env newState
-
--- EVALUATION
-
--- | Evaluate a @Value@ to Integer
+fixInterval interval state =
+    case interval of
+        (low, high)
+          | high < low -> IntervalError (InvalidInterval interval)
+          | otherwise -> let
+            curMinSlot = minSlot state
+            -- newLow is both new "low" and new "minSlot" (the lower bound for slotNum)
+            newLow = max low curMinSlot
+            -- We know high is greater or equal than newLow (prove)
+            curInterval = (newLow, high)
+            env = Environment { slotInterval = curInterval }
+            newState = state { minSlot = newLow }
+            in if high < curMinSlot then IntervalError (IntervalInPastError curMinSlot interval)
+            else IntervalTrimmed env newState
+
+
+{-|
+  Evaluates @Value@ given current @State@ and @Environment@
+-}
 evalValue :: Environment -> State -> Value -> Integer
```

Comment:
Why not produce `Ada`?

Reply:
Values can represent other things/quantities (e.g. slot numbers, choices), which could be compared in `Observation`.
Imagine ```If (SlotIntervalStart `ValueLT` ChoiceValue id def)```